types/ember-data/index.d.ts

```diff
@@ -546,11 +558,11 @@ export namespace DS {
         /**
          * Get the reference for the specified belongsTo relationship.
          */
-        belongsTo(name: RelationshipsFor<this>): BelongsToReference;
+        belongsTo<T extends keyof this = RelationshipsFor<this>>(name: T): BelongsToReference<UnpackComputedProperty<this[T]>>;
```

Comment:
Unfortunately `this` used as a typeparam has been a source of _significant breaking changes_  to our types as new compiler versions are released. The fact that it includes `protected` and `private` members, when used internally, but those _should not be visible externally_ results in some significant nuance that few other type packages seem to encounter (i.e., the ember types have been one of only 5-6 packages that break routinely around compiler changes that affect `this`)

As the person who often has to urgently dig us out of these kinds of problems, I'd prefer to 

(a) not use `<this>` or
(b) be ready to, at a moments notice, eject out to a _less clever_ typing in the event a new TS nightly build starts to block our CI pipeline

Reply:
What do you think about removing the `this` usage for belongsTo/hasMany and simply make HasManyReference and BelongsToReference generic?

I also hit some issues with the old `RelationshipsFor<this>` 

e.g.
```
belongsTo<T extends DS.Model>(name: string): BelongsToReference<T>;
hasMany<T extends DS.Model>(name: string): HasManyReference<T>;
```

Reply:
I'd consider this a possible misuse of generics. Generally they're intended to be used as a tool for expressing a relationship between two or more types (expressed in terms of one or more typeparams), and in the case above you use it in only once.

What you have here is just a casting mechanism where we'd have to provide a typeparam at every usage site. If we're going to encourage this kind of casting, I'd prefer that it _looks like_ casting (i.e., using the `as` keyword)

From a type safety standpoint,
```ts
belongsTo<T extends DS.Model>(name: string): BelongsToReference<T>;

this.belongsTo<Book>('book');
```
is no more type safe than 
```ts
belongsTo(name: string): BelongsToReference<DS.Model>;

this.belongsTo('book') as Book;
```

Reply:
Yes that makes sense. 

For us there were multiple places where the current api is really verbose if one wants to type it. e.g. the following is needed right now:

```
const ref = model.belongsTo('rel') 
ref.load() as Promise<Rel | null>;
ref.value() as Rel | null;
ref.reload() as Promise<Rel | null>;
```

Which could be changed to:
```
const ref = model.belongsTo('rel') as BelongsToRef<Rel>;
ref.load() // Promise<Rel | null>;
ref.value() // Rel | null;
ref.reload() // Promise<Rel | null>;
```

Which would be possible if we make `BelongsToReference` generic in the same way `HasManyReference` is currently generic. 

Would you be ok with that?